libeos-updater-util/flatpak.c

```diff
@@ -1457,6 +1457,651 @@ euu_filter_for_existing_flatpak_ref_actions (GHashTable *ref_actions  /* (elemen
+   * each remote to a list of refs, since we need to maintain the priority
+   * order. */
+  *out_remotes_priority_order = g_steal_pointer (&remotes);
+  *out_refs_for_remotes = g_steal_pointer (&refs_for_remotes);
+
+  return TRUE;
+}
+
+typedef struct _EuuFlatpakRelatedRefsForRemote {
+  FlatpakRemote *remote;
+  GHashTable    *related_refs;
+} EuuFlatpakRelatedRefsForRemote;
+
+static EuuFlatpakRelatedRefsForRemote *
+euu_flatpak_related_refs_for_remote_new (FlatpakRemote *remote,
+                                         GHashTable    *related_refs)
+{
+  EuuFlatpakRelatedRefsForRemote *related_refs_for_remote = g_new0 (EuuFlatpakRelatedRefsForRemote, 1);
+
+  related_refs_for_remote->remote = g_object_ref (remote);
+  related_refs_for_remote->related_refs = g_hash_table_ref (related_refs);
+
+  return related_refs_for_remote;
+}
+
+static void
+euu_flatpak_related_refs_for_remote_free (EuuFlatpakRelatedRefsForRemote *related_refs_for_remote)
+{
+  g_clear_object (&related_refs_for_remote->remote);
+  g_clear_pointer (&related_refs_for_remote->related_refs, (GDestroyNotify) g_hash_table_unref);
+
+  g_free (related_refs_for_remote);
+}
+
+/* The runtime ref is a special case that we cannot get from
+ * flatpak_installation_fetch_remote_metadata_sync. We need to manually
+ * fetch the metadata for each remote (this will usually come from the cache)
+ * and parse it for the runtime ref. */
+static gboolean
+fetch_remote_runtime_ref (FlatpakInstallation *installation,
+                          const gchar         *remote_name,
+                          FlatpakRef          *source_ref,
+                          GCancellable        *cancellable,
+                          FlatpakRemoteRef   **out_remote_runtime_ref,
+                          GError             **error)
+{
+  g_autoptr(FlatpakRef) runtime_ref = NULL;
+  g_autoptr(FlatpakRemoteRef) runtime_remote_ref = NULL;
+  g_autoptr(GBytes) bytes = NULL;
+  g_autoptr(GKeyFile) metadata = NULL;
+  g_autofree gchar *runtime_ref_str = NULL;
+  g_autofree gchar *runtime_ref_str_complete = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  g_return_val_if_fail (out_remote_runtime_ref != NULL, FALSE);
+
+  bytes = flatpak_installation_fetch_remote_metadata_sync (installation,
+                                                           remote_name,
+                                                           source_ref,
+                                                           cancellable,
+                                                           &local_error);
+
+  if (bytes == NULL)
+    {
+      /* G_IO_ERROR_NOT_FOUND here means that either the remote did not have
+       * a summary cache or that the source ref was not found in the
+       * summary cache. In either case, we just treat that as a non-fatal
+       * error and continue. */
+      if (g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        {
+          *out_remote_runtime_ref = NULL;
```

Comment:
Maybe add a `g_debug()` here before returning, so it’s easier to work out the control flow from debug logs.